### PR TITLE
Add named parameters for macros. 

### PR DESCRIPTION
**Pull request by ricbit**
Macros with lots of parameters can often become illegible.
For example, in "RECT 1, 2, 3, 4", what is the meaning of the numbers?
Instead of checking the definition of the macro, you can use named parameters to self-document the intent: "RECT x=1,y=2,dx=3,dy=4".
This PR implements the named parameters and add tests for it.

## Summary by CodeRabbit

* **New Features**
  * Macros support named-argument syntax (name = value), usable in any order; spaces around = accepted; mixing positional and named arguments is disallowed.

* **Bug Fixes**
  * Error messages now reliably show the correct macro name during parsing and arity errors.

* **Documentation**
  * User docs updated with named-argument syntax, rules, and examples.

* **Tests**
  * Added comprehensive tests covering valid named usage and many error cases.

* **Chores**
  * CI, coverage, and build scripts updated for improved workflows and coverage reporting.

### CODE REVIEW

File: tests/macros/named_macro_args.asm [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    ; spaces around '=' are allowed
+    LOAD_XY  x = 8 ,  y = 9
+
+    ; single argument macro
+    MACRO ONE val
+        DB val
+    ENDM
+
+    ONE val=10
```
Comment (ped7g) [inline tests/macros/named_macro_args.asm:24]: try also:
```
val: ONE <val=$>
```
this should pass with value `val=$` ie. `DB val=$` ie. `DB -1` -> `0xFF`.


And not sure what the valid syntax should be for named variant:
```
  ONE <val=val=$> ; everything is content, no named annotation here
  ONE val=<val=$> ; I guess this should work as `DB val=$`
```
I guess this one should pass as valid, `val=` is just name pinning annotation and `<...>` is valid  delimited argument content.

Weird one:
```
  ONE val=val=$
```
Not sure about this one, technically after removing name of arg rest `val=$` is content, so if your solution will accept this and end up with `DB val=$`, I'm ok with it. But if you will report this as error and require delimited content in `<>`, it feels consistent as some kind of rule (*"to have `=` in argument content, use delimited way"*)

In any case, please document these in test how they work or not, so we can discuss particular edge cases, what already works and what may be worth of change.

File: tests/macros/named_macro_args.asm [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        DB s
+    ENDM
+
+    TEXT s=<"a=b">
+
+    ; --- error cases ---
+
+    LOAD_XY x=1, z=2        ; unknown argument name
+    LOAD_XY x=1, x=2        ; duplicate argument name
+    LOAD_XY x=1             ; missing argument 'y'
```
Comment (ped7g) [inline tests/macros/named_macro_args.asm:37]: what about `LOAD_XY x=1, 2` and `LOAD_XY 2,x=1`?
I think in some languages once you use one name annotation, you are required to use all of them, in other languages the first would do implicit y=2, I hope the second variant doesn't work anywhere, because that's cursed.

What is your opinion + document it in test how it works for these cases where amount of args is correct, but explicit/implicit naming is mixed (I don't have preference right now, I would go with simpler implementation either supporting mixed way or error out, requiring all of them named is maybe slightly more "consistent" and less error prone, but implementation must be sane).

**Review by ped7g — COMMENTED** [review]
PR noted, but please ton of patience.

a) migration of CI is not fully finished and I will ping you back and request rebase after it is
b) I was hoping to finalize attempt to do .deb packaging before any bigger change (then enforced CI migration happened), so I may prefer to land this after .deb (or before ... depends how things will go with either thing).
c) meanwhile any attempt at updating documentation.xml would be nice addition to this PR
d) this theoretically breaks old valid syntax like `db_macro x=$`. But I don't believe anyone is using such macro argument value in live project, so I'm ok with changing syntax rules to "needs `<>` when content has `=` char", but should be documented and the delimited content inside  `<>` must work of course (seems it does, although you have also quotes there in the test, so it's not 100% clear).

Also please review: #72 #103 #150
and consider those ideas vs your change, if it happens to collide with anything and how, I think none of those should be affected by this, but reviewing it and being sure is better than just wild guess.

The syntax/rules should be clear and agreed before I would even take a look at the code itself, not reviewing that yet (plus pending CI and .deb issues being my current focus).

(obviously other maintainer of sj+ may step in at any time, but right now I'm only active one, so I'm a bit in a position of benevolent dictator, so have patience with me :D )

File: tests/macros/named_macro_args.asm (hunk 2 above)
Comment (ricbit) [reply tests/macros/named_macro_args.asm:37]: Both cases should be errors, I think. Once one parameter is named, all should be named.

File: tests/macros/named_macro_args.asm (hunk 1 above)
Comment (ricbit) [reply tests/macros/named_macro_args.asm:24]: What do you think about using := for named variables?
LOAD_XY x:=1, y:=2
I believe this will keep compatibility with the usage of = as the equal operator.
(I will refrain from updating the PR until we have a good syntax.)

**Comment by ricbit** [comment]
Regarding the open issues:

https://github.com/z00m128/sjasmplus/issues/72 
Macro Assembler syntax for named parameters is exactly the same as the proposed.
Sjasm doesnt have it, but the syntax is compatible.

https://github.com/z00m128/sjasmplus/issues/103 
Default parameters could work with named, no problem, compatible syntax.

https://github.com/z00m128/sjasmplus/issues/150
Types are an interesting idea, I also see no breaking syntax problem.

File: tests/macros/named_macro_args.asm (hunk 1 above)
Comment (ped7g) [reply tests/macros/named_macro_args.asm:24]: No, `:=` has two issues:

1) `:` splits lines (well before parser, at stream reading level, so the whole logic of detecting you are trying to call macro and not splitting line would have to move somewhere where it does not belong.

2) no matter what you chose for the syntax, it will always collide with old syntax, because in previous versions whatever-you-picked is sent as argument content.

ie. you chose for example `@@@=@@@` and do `ONE val@@@=@@@10` and the old sjasmplus will start executing body of macro `ONE` with `val` content equal to `val@@@=@@@10`.
The new version will then have `val` content equal `10`.

Whether you may have ever needed `@@@=@@@` in old source is not that important, because the need of `=` is already highly unlikely, probably nobody ever wrote that as macro input.

So let's accept there is clash and this is "breaking" change and rather focus to make it convenient to use and to not blatantly block possible future extensions (mentioned extra issues with ideas about macro arguments).

So `x=123` syntax is OK for me, I don't think you can do better.

**Comment by ped7g** [comment]
Ad CI migration: ok, I hope I'm done with it.

If you would patch the test to reflect the discussion and rebase on current master, it would be nice.
(no hurry, the .deb thing can take weeks ... and my spare time for OSS is limited, so even if **I** would hurry to get this finished, it will be days or weeks, not hours)

File: tests/macros/named_macro_args.asm (hunk 1 above)
Comment (ricbit) [reply tests/macros/named_macro_args.asm:24]: > try also:
> 
> ```
> val: ONE <val=$>
> ```
> 
> ```
>   ONE <val=val=$> ; everything is content, no named annotation here
> ```

On the current version, before the named params patch, both these cases lead to infinite recursion (eventually blocked by 31-limit or line too long).

I would propose that using the name of the param as a token when calling the macro should be an error. So your four ONE examples should be errors, because you are passing "val" and in macro with param name "val". Do you agree with this? Otherwise we would need to solve the infinite recursion problem before.

Comment (ped7g) [reply tests/macros/named_macro_args.asm:24]: Oh, I forgot about that, that's excellent news, so there's zero chance of `val=` as content in legacy sources, because it would cause error in assembling. So actually your initial suggestion doesn't even clash with old syntax (by side-effect of substitution rules causing infinite loop).

I think this is then OK as is, ie. `ONE val=$` will AFTER your change works because it's named argument, and `ONE <val=$>` will end in infinite recursion as before (confirm by test). The infinite recursion is good-enough error, no need for specific detection of name of param, the substitution detects it by loop. :)

So I would just add this to test to showcase the current behavior (inf. loop). If I did misunderstood something again, I will understand it from test, so that's the best way to tell me what happens. :)

Comment (ricbit) [reply tests/macros/named_macro_args.asm:24]: I have added new test cases and updated documentation, I think this is now good to go.

**Comment by ped7g** [comment]
Can you *rebase* the PR on latest master? If you are not familiar with rebasing, then it's not super important, I can do it myself when I will work on this PR, but it looks nice in history.

Also if/while rebasing, please edit the commit [b577ff8](https://github.com/z00m128/sjasmplus/pull/339/commits/b577ff8f366d6b379fbbdf5760b08331a24cdf63) to contain only .xml changes, don't commit updated .html, I'm updating .html only before release, so the github web online docs are for last released version, and .xml contains WIP changes.

(or let me know you don't want to try and I will eventually rebase+edit this myself)

And I can update you on current progress:
- CI is migrated
- v1.23.1 released and ready for Debian packaging attempt

Now I need to find some official Debian Developer to become a sponsor of the package and review it, hopefully this will happen within couple of weeks/months. Meanwhile I don't have any big plans to develop sjasmplus itself, so this PR may be stuck here for some time.

If the Debian adventure would take too long without clear progress, I will get back to this, otherwise I want to focus on .deb release, if it will be progressing well.

**Comment by ped7g** [comment]
ok, I'm looking into this from now on, don't add further edits (as I'm rebasing + editing the PR on my local machine, so the sources will not match anymore).

If you want to change something, add comment explaining what you want to change.

But generally speaking I like the idea and thank you for contribution. I don't like the implementation as much, but I need more time to review it thoroughly and decide what I will do about it, if I want to change it or not... In either case, it's excellent starting point, thank you.

**Comment by ped7g** [comment]
Moved to #363
